malta: enable kernel 5.10 and update configs

guidosarducci opened the pull request:
**Description**:
Allow configuring linux 5.10 as a testing kernel for the malta target, based on the 5.4 config with
additional kernel options needed to avoid prompts during build. Also prune subtarget configs of
auto-generated options.


**Testing**:
- Compile, boot, and run on QEMU using linux 5.10.16 for malta/mips{32,64}{be,le} archs
- Compile, boot, and run on QEMU using linux 5.4.99 for malta/mips32be and mips64le archs (regression test)
- Functional tested on all archs using `tc`, `ip`, various `kmod-sched-*` modules, and simple BPF programs

Commits could be squashed if needed too...

Thanks for any feedback!

guidosarducci committed:
malta: support kernel 5.10

Add linux 5.10 as testing kernel and copy generic config from 5.4.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
malta: update kernel 5.10 configs

Update generic config to add missing options that prompt during builds. Run
"make kernel_oldconfig" to refresh malta target config.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

neheb commented on target/linux/malta/config-5.10:
```diff
@@ -0,0 +1,340 @@
+CONFIG_ARCH_32BIT_OFF_T=y
```
This is interesting. What does it do?

guidosarducci replied:
It's automatically set from [arch/mips/Kconfig](https://github.com/torvalds/linux/blob/f40ddce88593482919761f74910f42f4b84c004b/arch/mips/Kconfig#L5). AFAICT it's used on legacy archs to [support 32-bit userspace offsets](https://github.com/torvalds/linux/blob/0653161f0faca68b77b3f36fb4b4b9b8b07050e5/arch/Kconfig#L320). This instance is simply copied from config-5.4.

guidosarducci commented:
Rebased to reflect some options being added in 0621b23efb4f88638052a6465cc8e96a6eb7d5e0 already. @blocktrron Could you have a look at this please?

guidosarducci commented on target/linux/malta/config-5.10:
```diff
@@ -0,0 +1,341 @@
+# CONFIG_CPU_MIPS32_R6 is not set
+# CONFIG_CPU_MIPS64_R1 is not set
+# CONFIG_CPU_MIPS64_R2 is not set
+# CONFIG_CPU_MIPS32_R5_FEATURES is not set
```
Thanks, will change.

guidosarducci committed:
malta: prune subtarget configs

Drop autogenerated kernel config options. These aren't needed to build and
can obscure options relevant to the subtarget.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci commented:
Updated per comments and rebased. Also pruned subtarget configs and re-tested.

guidosarducci commented:
Rebased against current master with it's kernel config updates and re-confirmed compile/run on QEMU/malta32be.

guidosarducci commented:
@hauke As you've done a fair bit of past maintenance on malta, can I ask you look at this update too? Thanks!

adschm commented:
Hmm, do you know about 'make kernel_oldconfig'?

guidosarducci commented:
You mean that command that's applied too liberally, often blindly, and sometimes bloats, obscures or butchers configs? Yes, I know about it. :-/  It was doing some strange things on this target, and I'm still in the middle of tracking down what's happening.

hauke commented:
You should run ```make kernel_oldconfig``` to refresh the main configuration when selecting one of the 32 bit subtargets and then for each subtarget run this: ```make kernel_oldconfig CONFIG_TARGET=subtarget```

guidosarducci commented:
@hauke I do run `make kernel_oldconfig` as a validation check; however, using the results reintroduces many non-prompt options that hurt readability and can confuse those looking at target/subtarget configs. For example, [this current change](https://github.com/openwrt/openwrt/pull/3881/files#diff-d8a1f200758ab6866c5dfd246e80984451cff299bbb249f9d1a28de1b2bc726a) will be completely reverted following your advice. I wont belabour the point, as that's been fruitless in the past.

So would you please clarify that refreshing is a prerequisite for merging, irrespective of the downsides? There's no issue updating if that's the case, but practices vary from member to member with no real documented guidance I've seen. Thanks.

guidosarducci commented:
@hauke Rebased and pushed an update refreshing the target kernel config from "make kernel_oldconfig". Biggest change is from the recent additions to `config-filter`. The series can be squashed if needed; it's split out only for review.

guidosarducci commented:
Rebased against current master with recent updates in generic 5.10 config.

neheb commented:
I seem to have this issue:

```
Package kmod-mdio-devres is missing dependencies for the following libraries:
of_mdio.ko
```

No idea if related to this PR.

neheb commented:
Never mind

```
--- a/package/kernel/linux/modules/netdevices.mk
+++ b/package/kernel/linux/modules/netdevices.mk
@@ -141,7 +141,7 @@ $(eval $(call KernelPackage,mii))
 define KernelPackage/mdio-devres
   SUBMENU:=$(NETWORK_DEVICES_MENU)
   TITLE:=Supports MDIO device registration
-  DEPENDS:=@LINUX_5_10 +kmod-libphy
+  DEPENDS:=@LINUX_5_10 +kmod-of-mdio +kmod-libphy
   KCONFIG:=CONFIG_MDIO_DEVRES
   HIDDEN:=1
   FILES:=$(LINUX_DIR)/drivers/net/phy/mdio_devres.ko
--- a/package/kernel/linux/modules/usb.mk
+++ b/package/kernel/linux/modules/usb.mk
@@ -1216,6 +1216,7 @@ $(eval $(call KernelPackage,usb-net-mcs7830))

 define KernelPackage/usb-net-smsc95xx
   TITLE:=SMSC LAN95XX based USB 2.0 10/100 ethernet devices
+  DEPENDS:=+kmod-libphy
   KCONFIG:=CONFIG_USB_NET_SMSC95XX
   FILES:=$(LINUX_DIR)/drivers/$(USBNET_DIR)/smsc95xx.ko
   AUTOLOAD:=$(call AutoProbe,smsc95xx)
```

guidosarducci commented:
> I seem to have this issue:
> 
> ```
> Package kmod-mdio-devres is missing dependencies for the following libraries:
> of_mdio.ko
> ```
> 
> No idea if related to this PR.

@neheb Not this PR, but I added `kmod-mdio-devres` as part of #3989, so better to post there.

I tested with some `CONFIG_ALL_KMODS` builds and didn't need the `of_mdio.ko` dependency. Can you post your diffconfig for that error?

Looking at the source, I think the dependency should be something like:

`DEPENDS:=@LINUX_5_10 +kmod-libphy PACKAGE_kmod-of-mdio:kmod-of-mdio`

i.e. the dependency is there only if CONFIG_OF_MDIO from the package is defined elsewhere.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

guidosarducci commented:
@ynezz Thank you, Petr.

I'd be grateful if you could also share thoughts on https://github.com/openwrt/openwrt/pull/3989#issuecomment-799226623. I really need to confirm a preference from devs between adding an `elfutils` host build and just setting a new build prerequisite (with potential roll-out problems on buildbots).
